File: wechat/miniprogram/api/totp.ts
import { CODE } from "@constant/error";
import { PATH } from "@constant/totp";
import { HttpError } from "@models/error";
import error from "@utils/error";
import http from "@utils/http";
import logger from "@utils/logger";
import type {
  CreateRequest,
  DeleteRequest,
  DetailRequest,
  EditIssuerRequest,
  EditUsernameRequest,
  Item,
} from "types/totp";

const all = async () => {
  try {
    return await http.post<Item[]>(PATH.ALL);
  } catch (e: unknown) {
    logger.error("查询 TOTP 列表失败", e);

    throw new HttpError(CODE.HTTP_API_TOTP_ALL, error.getErrorMessage(e));
  }
};

const detail = async (id: string) => {
  try {
    return await http.post<Item>(PATH.DETAIL, { id } as DetailRequest);
  } catch (e: unknown) {
    logger.error("查询 TOTP 详情失败", e);

    throw new HttpError(CODE.HTTP_API_TOTP_DETAIL, error.getErrorMessage(e));
  }
};

const create = async (uri: string) => {
  try {
    return await http.post<null>(PATH.CREATE, { uri } as CreateRequest);
  } catch (e: unknown) {
    logger.error("创建 TOTP 失败", e);

    throw new HttpError(CODE.HTTP_API_TOTP_CREATE, error.getErrorMessage(e));
  }
};

const editIssuer = async (data: EditIssuerRequest) => {
  try {
    return await http.post<null>(PATH.EDIT_ISSUER, data);
  } catch (e: unknown) {
    logger.error("更新 TOTP 的 Issuer 信息失败", e);

    throw new HttpError(
      CODE.HTTP_API_TOTP_EDIT_ISSUER,
      error.getErrorMessage(e),
    );
  }
};

const editUsername = async (data: EditUsernameRequest) => {
  try {
    return await http.post<null>(PATH.EDIT_USERNAME, data);
  } catch (e: unknown) {
    logger.error("更新 TOTP 的 username 信息失败", e);

    throw new HttpError(
      CODE.HTTP_API_TOTP_EDIT_USERNAME,
      error.getErrorMessage(e),
    );
  }
};

const deleteTotp = async (id: string) => {
  try {
    return await http.post<null>(PATH.DELETE, { id } as DeleteRequest);
  } catch (e: unknown) {
    logger.error("删除 TOTP 失败", e);

    throw new HttpError(CODE.HTTP_API_TOTP_ALL, error.getErrorMessage(e));
  }
};

export default { all, detail, create, editIssuer, editUsername, deleteTotp };

File: application-api/src/v1/totp.rs
use axum::Extension;

use crate::extract::Json;
use crate::request::Validator;
use crate::request::totp::{
    CreateRequest, DeleteRequest, DetailRequest, DetailResponse, EditIssuerRequest,
    EditUsernameRequest,
};
use crate::response::Resp;
use crate::response::Response;
use crate::service;
use application_database::account::access_token::AccessToken;

pub async fn all(Extension(access_token): Extension<AccessToken>) -> Resp<Vec<DetailResponse>> {
    Ok(Response::success(service::totp::all(access_token).await?))
}

pub async fn detail(
    Extension(access_token): Extension<AccessToken>,
    Json(request): Json<DetailRequest>,
) -> Resp<DetailResponse> {
    let id = request.validate()?;

    Ok(Response::success(
        service::totp::detail(access_token, id).await?,
    ))
}

pub async fn create(
    Extension(access_token): Extension<AccessToken>,
    Json(request): Json<CreateRequest>,
) -> Resp<()> {
    let uri = request.validate()?;

    service::totp::create(access_token, uri).await?;

    Ok(Response::success(()))
}

pub async fn edit_issuer(
    Extension(access_token): Extension<AccessToken>,
    Json(request): Json<EditIssuerRequest>,
) -> Resp<()> {
    let params = request.validate()?;

    service::totp::edit_issuer(access_token, params).await?;

    Ok(Response::success(()))
}

pub async fn edit_username(
    Extension(access_token): Extension<AccessToken>,
    Json(request): Json<EditUsernameRequest>,
) -> Resp<()> {
    let params = request.validate()?;

    service::totp::edit_username(access_token, params).await?;

    Ok(Response::success(()))
}

pub async fn delete(
    Extension(access_token): Extension<AccessToken>,
    Json(request): Json<DeleteRequest>,
) -> Resp<()> {
    let id = request.validate()?;

    service::totp::delete(access_token, id).await?;

    Ok(Response::success(()))
}
